aws/request_pagination_test.go

```diff
@@ -513,152 +513,93 @@ func TestPaginationWithContextNilInput(t *testing.T) {
 	type testPageOutput struct {
-		Value     string
+		Value     *string
 		NextToken *string
 	}
+	type testCase struct {
+		Value, PrevToken, NextToken *string
+	}
 
-	expect := []struct {
-		Value, PrevToken, NextToken string
-	}{
-		{"FirstValue", "InitalToken", "FirstToken"},
-		{"SecondValue", "FirstToken", "SecondToken"},
-		{"ThirdValue", "SecondToken", ""},
-	}
-	input := testPageInput{
-		NextToken: expect[0].PrevToken,
-	}
-
-	c := awstesting.NewClient(unit.Config())
-	i := 0
-	p := aws.Pagination{
-		NewRequest: func() (*aws.Request, error) {
-			r := c.NewRequest(
-				&aws.Operation{
-					Name: "Operation",
-					Paginator: &aws.Paginator{
-						InputTokens:  []string{"NextToken"},
-						OutputTokens: []string{"NextToken"},
-					},
-				},
-				&input, &testPageOutput{},
-			)
-			// Setup handlers for testing
-			r.Handlers.Clear()
-			r.Handlers.Build.PushBack(func(req *aws.Request) {
-				in := req.Params.(*testPageInput)
-				if e, a := expect[i].PrevToken, in.NextToken; e != a {
-					t.Errorf("%d, expect NextToken input %q, got %q", i, e, a)
-				}
-			})
-			r.Handlers.Unmarshal.PushBack(func(req *aws.Request) {
-				out := &testPageOutput{
-					Value: expect[i].Value,
-				}
-				if len(expect[i].NextToken) > 0 {
-					out.NextToken = aws.String(expect[i].NextToken)
-				}
-				req.Data = out
-			})
-			return r, nil
+	cases := map[string][]testCase{
+		"nil NextToken": {
+			testCase{aws.String("FirstValue"), aws.String("InitalToken"), aws.String("FirstToken")},
+			testCase{aws.String("SecondValue"), aws.String("FirstToken"), aws.String("SecondToken")},
+			testCase{aws.String("ThirdValue"), aws.String("SecondToken"), nil},
+		},
+		"zero NextToken": {
+			testCase{aws.String("FirstValue"), aws.String("InitalToken"), aws.String("FirstToken")},
+			testCase{aws.String("SecondValue"), aws.String("FirstToken"), aws.String("SecondToken")},
+			testCase{aws.String("ThirdValue"), aws.String("SecondToken"), aws.String("")},
 		},
 	}
 
-	for p.Next() {
-		data := p.Page().(*testPageOutput)
+	for cName, c := range cases {
+		t.Run(cName, func(t *testing.T) {
+			input := testPageInput{
+				NextToken: c[0].PrevToken,
+			}
 
-		if e, a := expect[i].Value, data.Value; e != a {
-			t.Errorf("%d, expect Value to be %q, got %q", i, e, a)
-		}
-		if e, a := expect[i].NextToken, aws.StringValue(data.NextToken); e != a {
-			t.Errorf("%d, expect NextToken to be %q, got %q", i, e, a)
-		}
+			svc := awstesting.NewClient(unit.Config())
+			i := 0
+			p := aws.Pagination{
+				NewRequest: func() (*aws.Request, error) {
+					r := svc.NewRequest(
+						&aws.Operation{
+							Name: "Operation",
+							Paginator: &aws.Paginator{
+								InputTokens:  []string{"NextToken"},
+								OutputTokens: []string{"NextToken"},
+							},
+						},
+						&input, &testPageOutput{},
+					)
+					// Setup handlers for testing
+					r.Handlers.Clear()
+					r.Handlers.Build.PushBack(func(req *aws.Request) {
+						if e, a := len(c), i+1; a > e {
```

Comment:
this check seems weird. Wouldn't e always be greater than a? Obviously missing something as this test passes.

The first test case e would be three. Calling the first `Next` would result in `i` being `0`. So would this check fail??

Reply:
This check is to identify if too many requests have been made and fail the build, because `i` now is greater than the number of expected tests.

This could be simplified as just `e, a := len(c), i; a == e`  Using equal instead of greater than would remove the need for the `+1`